### PR TITLE
feat: add exception handling

### PR DESCRIPTION
**Pull request by emizzle**
Closes: #239

Remove all `except Exception as e` where possible. In some instances, transitive dependencies out of our control were the source of this error. Even in instances of the Nim library, like `parseJson`, for example, force us to `except Exception` to prevent it from bubbling up to higher layers in the abstraction.

At the lowest levels, we are catching all errors that inherit from `CatchableError` and translating them to be useful for layers above. At the highest level, which is our public API layer, meant to be directly consumed by consumer clients, we are catching all catchable exceptions from the layers below and translating them in to a Result type.

Developer guid documentation for exception handling that explains the motivation for these changes in its own [developer guide](https://hackmd.io/rkpbKsEYT_umNlHU1IcbiQ)

### CODE REVIEW

File: status/api/accounts.nim
Comment (arnetheduck) [inline status/api/accounts.nim:128]: another small detail worth considering: exceptions and `Result` are both error handling mechanisms that by and large address the same problem - we tend to avoid using both in the same function, else the calling code must go through some gymnastics to handle both types - one case were this potentially doesn't apply is if there's some truly exceptional construct that _should_ ignore all "normal" error handling and jump straight to an exception handler, but that's .. rare.

Comment (emizzle) [reply status/api/accounts.nim:128]: Very good point and I did strive for that. Looks like this one slipped through. Will fix so that the StatusApiError is caught.

Comment (michaelsbradleyjr) [reply status/api/accounts.nim:128]: Making sure I understand... so if the bases are covered inside the proc, then we don't need/want the raises pragma here (keeping in mind that at the top of the module we have `{.push raises: [Defect].}`)?

Comment (arnetheduck) [reply status/api/accounts.nim:128]: > we don't need/want the raises pragma here

correct

Comment (emizzle) [reply status/api/accounts.nim:128]: I believe I did catch StatusApiError in other top-level, public-facing procs. Really think I just missed this one. May be others as well, will review.

**Comment by michaelsbradleyjr** [comment]
Tests were failing locally for me, just like in CI. I added a WIP commit that adds Defect to the raises pragma for initializeDB, and now tests are passing locally and in CI.

File: status/api/wallet.nim
Comment (michaelsbradleyjr) [inline status/api/wallet.nim:129]: Is it important for some procs to basically undo what `{.push raises: [Defect].}` does at the top of the module?

File: status/private/chats.nim
Comment (michaelsbradleyjr) [inline status/private/chats.nim:4]:
```suggestion
  std/[json, marshal, options, sequtils, strformat]
```

File: status/private/identicon.nim
Comment (michaelsbradleyjr) [inline status/private/identicon.nim:106]:
```suggestion
    raise (ref IdenticonError)(parent: e, msg: "Error generating base64 " &
      "encoded identicon")
```

File: status/private/migration.nim
Comment (michaelsbradleyjr) [inline status/private/migration.nim:28]:
```suggestion
  const errorMsg = "Error creating migration table if not exists"
```

File: status/private/tx_history.nim
Comment (michaelsbradleyjr) [inline status/private/tx_history.nim:586]: I realize this code didn't originate with this PR, but we can't `sleep`, as that would block the thread on which nim-status is running. It should be up to a client implementor (e.g. ourselves with the example client) to fetch history in a loop using `sleepAsync`.

Also, as seen later in this module, this logic involves a lock and and an extra thread, which is also a no-go. I'ved created an issue re: this module needing to be overhauled: #266.

**Review by michaelsbradleyjr — COMMENTED** [review]
Awesome work, what a monumental effort!  I had to take breaks while reviewing it. 😅 

Below I have a couple of questions and some minor style-related adjustments. I also made some comments re: problems in `status/private/tx_history.nim`, but fixing those is definitely a separate task.

Note my comment above re: adding a WIP commit; should be squashed into the main commit before merge.

Also, given the huge scope of this PR, I'm sure there are things here and there that could be touched up and I didn't notice them. But the fact it compiles correctly, the tests all pass, and the example client compiles and runs correctly — that's all a testament to your fortitude and thorough work on it. Thank you for tackling this!

File: status/api/accounts.nim
Comment (emizzle) [reply status/api/accounts.nim:31]: Correct, it is not needed. I've removed from here and another instance as well. Likely this was leftover from the first step changes being in place with all possible exceptions listed, then slowly chipping away each except from the list until only defect was left, at which point I should have just removed the pragma altogether.

File: status/api/accounts.nim
Comment (emizzle) [reply status/api/accounts.nim:32]: Done.

File: status/api/accounts.nim
Comment (emizzle) [reply status/api/accounts.nim:65]: Agreed, updated.

File: status/api/wallet.nim
Comment (emizzle) [reply status/api/wallet.nim:129]: It means that there is no chance for that proc to raise a `Defect` based on the compiler's analysis. Because `{.push raises: [Defect].}` effectively adds a `{.raises: [Defect].}` to each proc/func/method/template/macro in the module, if there is a proc/func/method/template/macro that does not raise `Defect`, the `raises` pragma needs to be explicitly overridden for the compiler to understand.

File: status/private/chats.nim
Comment (emizzle) [reply status/private/chats.nim:4]: Good catch, updated.

File: status/private/identicon.nim
Comment (emizzle) [reply status/private/identicon.nim:106]: Updated.

File: status/private/migration.nim
Comment (emizzle) [reply status/private/migration.nim:28]: Great catch, updated.

File: status/private/tx_history.nim
Comment (emizzle) [reply status/private/tx_history.nim:586]: Definitely agreed. There are no code paths from our public APIs to this module, nor any calls internally to this code. As discussed, we should repurpose it to be consumed by the client in a manner suited to the architecture of the client. In the case of the example chat client, we could start a long-running task that does this work.

**Comment by emizzle** [comment]
> Note my comment above re: adding a WIP commit; should be squashed into the main commit before merge.

All squashed. Should be good to merge if approved.

File: status/api/wallet.nim
Comment (arnetheduck) [reply status/api/wallet.nim:129]: > It means that there is no chance for that proc to raise a Defect based on the compiler's analysis. 

This unfortunately is not quite so - in nim 1.4, `raises: []` means that `Defect` can still be raised, while in nim 1.2 this transition was done only partially and some `Defect` trigger a compile-time error and some don't - we use `push raises: [Defect]` to prepare the code for nim 1.4 where it'll be replaced with `push raises: []` throughout, because this is the semantic that most closely matches where nim is heading.

Comment (emizzle) [reply status/api/wallet.nim:129]: Makes sense Jacek, thanks so much for the info.

**Review by michaelsbradleyjr — APPROVED** [review]

**Pull request merged**
